File: packages/core/src/validation.ts
/**
 * Deterministic validation of an extracted Indian GST tax invoice.
 *
 * Nothing here estimates. Every check either holds or does not, which is why
 * these findings, and not a model's self-reported confidence, are the primary
 * input to review routing.
 *
 * Rule references are to the CGST Rules, 2017.
 */

import { isValidGstin } from "./gstin";
import { absolute, impliedRateBasisPoints, parsePaise } from "./money";
import { type Field, type FieldName, type InvoiceExtraction, isPresent } from "./model";

/**
 * Rounding slack, in paise. Amounts are rounded to two decimals per line and
 * per tax head, so a one-paisa drift is a rounding artefact, not an error.
 */
const TOLERANCE_PAISE = 1n;

/** Rule 46(b): at most sixteen characters, alphanumerics plus hyphen and slash. */
const INVOICE_NUMBER_MAX_LENGTH = 16;
const INVOICE_NUMBER_PATTERN = /^[A-Za-z0-9/-]+$/;

/** Rule 46(g): HSN for goods, SAC for services. Four, six or eight digits. */
const HSN_LENGTHS: ReadonlySet<number> = new Set([4, 6, 8]);

const ISO_DATE_PATTERN = /^\d{4}-\d{2}-\d{2}$/;

/**
 * Combined ad valorem rates in basis points, across the pre- and post-reform
 * slab structures. A plausibility heuristic, not a legal rule: an unrecognised
 * rate means look again, not reject.
 */
const PLAUSIBLE_RATE_BASIS_POINTS: ReadonlySet<bigint> = new Set([
  0n, 10n, 25n, 100n, 150n, 300n, 500n, 600n, 750n, 1200n, 1800n, 2800n, 4000n,
]);

/** Lower sorts first. FATAL means the document is not an invoice at all. */
export enum Severity {
  Fatal = 0,
  Error = 1,
  Warning = 2,
}

export interface Finding {
  readonly code: string;
  readonly severity: Severity;
  readonly field: FieldName;
  readonly message: string;
}

const MANDATORY_PARTICULARS: ReadonlyArray<readonly [FieldName, string]> = [
  ["supplierGstin", "MISSING_SUPPLIER_GSTIN"],
  ["invoiceNumber", "MISSING_INVOICE_NUMBER"],
  ["invoiceDate", "MISSING_INVOICE_DATE"],
  ["taxableValue", "MISSING_TAXABLE_VALUE"],
  ["totalValue", "MISSING_TOTAL_VALUE"],
];

const AMOUNT_FIELDS: ReadonlyArray<readonly [FieldName, string]> = [
  ["taxableValue", "TAXABLE_VALUE_NOT_NUMERIC"],
  ["totalValue", "TOTAL_VALUE_NOT_NUMERIC"],
  ["cgstAmount", "CGST_AMOUNT_NOT_NUMERIC"],
  ["sgstAmount", "SGST_AMOUNT_NOT_NUMERIC"],
  ["igstAmount", "IGST_AMOUNT_NOT_NUMERIC"],
  ["cessAmount", "CESS_AMOUNT_NOT_NUMERIC"],
];

/** Every deterministic rule this extraction violates, most severe first. */
export function validate(extraction: InvoiceExtraction): Finding[] {
  const findings: Finding[] = [
    ...checkMandatoryParticulars(extraction),
    ...checkInvoiceNumber(extraction.invoiceNumber),
    ...checkGstins(extraction),
    ...checkHsn(extraction.hsn),
    ...checkInvoiceDate(extraction.invoiceDate),
    ...checkArithmetic(extraction),
    ...checkTaxHeads(extraction),
  ];
  return findings.sort((a, b) => a.severity - b.severity);
}

/**
 * Rule 46: without these the document is not a tax invoice. The recipient's
 * GSTIN is absent from this list because Rule 46(d) requires it only where the
 * recipient is registered.
 */
function checkMandatoryParticulars(extraction: InvoiceExtraction): Finding[] {
  return MANDATORY_PARTICULARS.filter(([name]) => !isPresent(extraction[name])).map(
    ([name, code]) => ({
      code,
      severity: Severity.Fatal,
      field: name,
      message: `Rule 46 requires ${name}`,
    }),
  );
}

function checkInvoiceNumber(f: Field): Finding[] {
  if (!isPresent(f)) return [];
  const value = (f.value as string).trim();
  const findings: Finding[] = [];
  if (value.length > INVOICE_NUMBER_MAX_LENGTH) {
    findings.push({
      code: "INVOICE_NUMBER_TOO_LONG",
      severity: Severity.Error,
      field: "invoiceNumber",
      message: `Rule 46(b) caps the invoice number at ${INVOICE_NUMBER_MAX_LENGTH} characters; got ${value.length}`,
    });
  }
  if (!INVOICE_NUMBER_PATTERN.test(value)) {
    findings.push({
      code: "INVOICE_NUMBER_BAD_CHARS",
      severity: Severity.Error,
      field: "invoiceNumber",
      message: "Rule 46(b) permits alphanumerics, hyphen and slash only",
    });
  }
  return findings;
}

function checkGstins(extraction: InvoiceExtraction): Finding[] {
  const pairs: ReadonlyArray<readonly [FieldName, string]> = [
    ["supplierGstin", "SUPPLIER_GSTIN_INVALID"],
    ["recipientGstin", "RECIPIENT_GSTIN_INVALID"],
  ];
  return pairs
    .filter(([name]) => isPresent(extraction[name]) && !isValidGstin(extraction[name].value))
    .map(([name, code]) => ({
      code,
      severity: Severity.Error,
      field: name,
      message: "GSTIN failed structural or check-digit verification",
    }));
}

function checkHsn(f: Field): Finding[] {
  if (!isPresent(f)) return [];
  const value = (f.value as string).trim();
  if (/^\d+$/.test(value) && HSN_LENGTHS.has(value.length)) return [];
  return [
    {
      code: "HSN_MALFORMED",
      severity: Severity.Error,
      field: "hsn",
      message: "Rule 46(g) HSN/SAC must be 4, 6 or 8 digits",
    },
  ];
}

/**
 * Dates must arrive already normalised to ISO 8601. Day-month ambiguity is
 * resolved upstream where the document's own format is visible; by the time a
 * date reaches validation an unparseable value means normalisation failed.
 */
function checkInvoiceDate(f: Field): Finding[] {
  if (!isPresent(f)) return [];
  const value = (f.value as string).trim();
  const malformed: Finding = {
    code: "INVOICE_DATE_UNPARSEABLE",
    severity: Severity.Error,
    field: "invoiceDate",
    message: "invoice date is not a normalised ISO 8601 date",
  };
  if (!ISO_DATE_PATTERN.test(value)) return [malformed];
  // Date.parse accepts 2026-02-31 and rolls it forward, so round-trip instead.
  const parsed = new Date(`${value}T00:00:00Z`);
  if (Number.isNaN(parsed.getTime())) return [malformed];
  return parsed.toISOString().slice(0, 10) === value ? [] : [malformed];
}

function checkArithmetic(extraction: InvoiceExtraction): Finding[] {
  const findings: Finding[] = [];

  for (const [name, code] of AMOUNT_FIELDS) {
    const f = extraction[name];
    if (isPresent(f) && parsePaise(f.value) === null) {
      findings.push({
        code,
        severity: Severity.Error,
        field: name,
        message: "amount is not a parseable decimal",
      });
    }
  }

  const amount = (name: FieldName): bigint | null =>
    isPresent(extraction[name]) ? parsePaise(extraction[name].value) : null;
  const zeroIfAbsent = (name: FieldName): bigint | null =>
    isPresent(extraction[name]) ? parsePaise(extraction[name].value) : 0n;

  const taxable = amount("taxableValue");
  const total = amount("totalValue");
  const cgst = zeroIfAbsent("cgstAmount");
  const sgst = zeroIfAbsent("sgstAmount");
  const igst = zeroIfAbsent("igstAmount");
  const cess = zeroIfAbsent("cessAmount");

  if (taxable !== null && taxable < 0n) {
    findings.push({
      code: "NEGATIVE_TAXABLE_VALUE",
      severity: Severity.Error,
      field: "taxableValue",
      message: "taxable value is negative; a credit note is a separate document type",
    });
  }

  if (taxable === null || total === null || cgst === null || sgst === null || igst === null || cess === null) {
    return findings;
  }

  const expectedTotal = taxable + cgst + sgst + igst + cess;
  if (absolute(expectedTotal - total) > TOLERANCE_PAISE) {
    findings.push({
      code: "TOTAL_MISMATCH",
      severity: Severity.Error,
      field: "totalValue",
      message: "taxable plus taxes does not equal the invoice total",
    });
  }

  if (cgst > 0n && sgst > 0n && absolute(cgst - sgst) > TOLERANCE_PAISE) {
    findings.push({
      code: "CGST_SGST_ASYMMETRY",
      severity: Severity.Error,
      field: "cgstAmount",
      message: "central and state tax must be equal halves of an intra-state levy",
    });
  }

  const rate = impliedRateBasisPoints(cgst + sgst + igst, taxable);
  if (rate !== null && !PLAUSIBLE_RATE_BASIS_POINTS.has(rate)) {
    findings.push({
      code: "IMPLAUSIBLE_TAX_RATE",
      severity: Severity.Warning,
      field: "taxableValue",
      message: `implied combined tax rate of ${Number(rate) / 100}% is not a GST slab`,
    });
  }

  return findings;
}

/**
 * Place of supply decides the tax head. Rule 46(l), (m) and (n). Supplier state
 * equal to place of supply means an intra-state supply taxed as CGST plus SGST;
 * otherwise it is inter-state and taxed as IGST. The head is never read off a
 * label on the page.
 */
function checkTaxHeads(extraction: InvoiceExtraction): Finding[] {
  const supplier = extraction.supplierGstin;
  const pos = extraction.placeOfSupplyStateCode;
  if (!isPresent(supplier) || !isValidGstin(supplier.value)) return [];
  if (!isPresent(pos)) return [];

  const cgst = parsePaise(extraction.cgstAmount.value) ?? 0n;
  const sgst = parsePaise(extraction.sgstAmount.value) ?? 0n;
  const igst = parsePaise(extraction.igstAmount.value) ?? 0n;

  const findings: Finding[] = [];
  if (igst > 0n && (cgst > 0n || sgst > 0n)) {
    findings.push({
      code: "MIXED_TAX_HEADS",
      severity: Severity.Error,
      field: "igstAmount",
      message: "a supply is either intra-state or inter-state, never both",
    });
  }

  const supplierState = (supplier.value as string).trim().slice(0, 2);
  const placeOfSupply = (pos.value as string).trim().padStart(2, "0");
  const isIntraState = supplierState === placeOfSupply;

  if (isIntraState && igst > 0n) {
    findings.push({
      code: "WRONG_TAX_HEAD",
      severity: Severity.Error,
      field: "igstAmount",
      message: "intra-state supply charged IGST",
    });
  }
  if (!isIntraState && (cgst > 0n || sgst > 0n)) {
    findings.push({
      code: "WRONG_TAX_HEAD",
      severity: Severity.Error,
      field: "cgstAmount",
      message: "inter-state supply charged CGST or SGST",
    });
  }
  return findings;
}

File: src/invoice_extract/validation.py
"""Deterministic validation of an extracted Indian GST tax invoice.

Nothing here estimates. Every check either holds or does not, which is why
these findings, and not a model's self-reported confidence score, are the
primary input to review routing.

Rule references are to the CGST Rules, 2017.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import date
from decimal import Decimal, InvalidOperation
from enum import IntEnum

from .gstin import is_valid_gstin
from .model import Field, InvoiceExtraction

# Rounding slack. Amounts are rounded to two decimals per line and per head,
# so a one-paisa drift is a rounding artefact, not an extraction error.
TOLERANCE = Decimal("0.01")

# Rule 46(b): not more than sixteen characters, alphanumerics plus hyphen and
# slash only, unique for a financial year.
INVOICE_NUMBER_MAX_LENGTH = 16
INVOICE_NUMBER_PATTERN = re.compile(r"^[A-Za-z0-9/-]+$")

# Rule 46(g): HSN for goods, SAC for services. Four, six or eight digits.
HSN_LENGTHS = frozenset({4, 6, 8})

# Combined ad valorem rates seen across the pre- and post-reform slab
# structures. This is a plausibility heuristic, not a legal rule: an
# unrecognised rate means look again, not reject.
PLAUSIBLE_TAX_RATES = frozenset(
    Decimal(x)
    for x in ("0", "0.1", "0.25", "1", "1.5", "3", "5", "6", "7.5", "12", "18", "28", "40")
)


class Severity(IntEnum):
    """Lower sorts first. FATAL means the document is not an invoice at all."""

    FATAL = 0
    ERROR = 1
    WARNING = 2


@dataclass(frozen=True)
class Finding:
    code: str
    severity: Severity
    field: str
    message: str


def _amount(field: Field) -> Decimal | None:
    """Parse an amount, or None if it is absent or not a number.

    Deliberately strict. A value like "Rs. 1,180/-" is an upstream
    normalisation failure and must surface as a finding, not be silently
    coerced into 1180.
    """
    if not field.is_present:
        return None
    try:
        return Decimal(str(field.value).strip())
    except (InvalidOperation, ValueError):
        return None


def _zero_if_absent(field: Field) -> Decimal | None:
    return Decimal("0") if not field.is_present else _amount(field)


def validate(extraction: InvoiceExtraction) -> list[Finding]:
    """Every deterministic rule this extraction violates, most severe first."""
    findings: list[Finding] = []
    findings.extend(_check_mandatory_particulars(extraction))
    findings.extend(_check_invoice_number(extraction.invoice_number))
    findings.extend(_check_gstins(extraction))
    findings.extend(_check_hsn(extraction.hsn))
    findings.extend(_check_invoice_date(extraction.invoice_date))
    findings.extend(_check_arithmetic(extraction))
    findings.extend(_check_tax_heads(extraction))
    return sorted(findings, key=lambda finding: finding.severity)


def _check_mandatory_particulars(extraction: InvoiceExtraction) -> list[Finding]:
    """Rule 46: without these the document is not a tax invoice.

    The recipient's GSTIN is not on this list because Rule 46(d) requires it
    only where the recipient is registered.
    """
    required = {
        "supplier_gstin": "MISSING_SUPPLIER_GSTIN",
        "invoice_number": "MISSING_INVOICE_NUMBER",
        "invoice_date": "MISSING_INVOICE_DATE",
        "taxable_value": "MISSING_TAXABLE_VALUE",
        "total_value": "MISSING_TOTAL_VALUE",
    }
    return [
        Finding(code, Severity.FATAL, name, f"Rule 46 requires {name.replace('_', ' ')}")
        for name, code in required.items()
        if not getattr(extraction, name).is_present
    ]


def _check_invoice_number(field: Field) -> list[Finding]:
    if not field.is_present:
        return []
    value = str(field.value).strip()
    findings = []
    if len(value) > INVOICE_NUMBER_MAX_LENGTH:
        findings.append(
            Finding(
                "INVOICE_NUMBER_TOO_LONG",
                Severity.ERROR,
                "invoice_number",
                f"Rule 46(b) caps the invoice number at {INVOICE_NUMBER_MAX_LENGTH} "
                f"characters; got {len(value)}",
            )
        )
    if not INVOICE_NUMBER_PATTERN.match(value):
        findings.append(
            Finding(
                "INVOICE_NUMBER_BAD_CHARS",
                Severity.ERROR,
                "invoice_number",
                "Rule 46(b) permits alphanumerics, hyphen and slash only",
            )
        )
    return findings


def _check_gstins(extraction: InvoiceExtraction) -> list[Finding]:
    findings = []
    for name, code in (
        ("supplier_gstin", "SUPPLIER_GSTIN_INVALID"),
        ("recipient_gstin", "RECIPIENT_GSTIN_INVALID"),
    ):
        field = getattr(extraction, name)
        if field.is_present and not is_valid_gstin(str(field.value)):
            findings.append(
                Finding(
                    code,
                    Severity.ERROR,
                    name,
                    "GSTIN failed structural or check-digit verification",
                )
            )
    return findings


def _check_hsn(field: Field) -> list[Finding]:
    if not field.is_present:
        return []
    value = str(field.value).strip()
    if value.isdigit() and len(value) in HSN_LENGTHS:
        return []
    return [
        Finding(
            "HSN_MALFORMED",
            Severity.ERROR,
            "hsn",
            "Rule 46(g) HSN/SAC must be 4, 6 or 8 digits",
        )
    ]


def _check_invoice_date(field: Field) -> list[Finding]:
    """Dates must arrive already normalised to ISO 8601.

    Day-month ambiguity is resolved upstream where the document's own format
    is visible; by the time a date reaches validation, an unparseable value
    means normalisation failed and the value cannot be trusted.
    """
    if not field.is_present:
        return []
    try:
        date.fromisoformat(str(field.value).strip())
    except ValueError:
        return [
            Finding(
                "INVOICE_DATE_UNPARSEABLE",
                Severity.ERROR,
                "invoice_date",
                "invoice date is not a normalised ISO 8601 date",
            )
        ]
    return []


def _check_arithmetic(extraction: InvoiceExtraction) -> list[Finding]:
    findings: list[Finding] = []

    for name, code in (
        ("taxable_value", "TAXABLE_VALUE_NOT_NUMERIC"),
        ("total_value", "TOTAL_VALUE_NOT_NUMERIC"),
        ("cgst_amount", "CGST_AMOUNT_NOT_NUMERIC"),
        ("sgst_amount", "SGST_AMOUNT_NOT_NUMERIC"),
        ("igst_amount", "IGST_AMOUNT_NOT_NUMERIC"),
        ("cess_amount", "CESS_AMOUNT_NOT_NUMERIC"),
    ):
        field = getattr(extraction, name)
        if field.is_present and _amount(field) is None:
            findings.append(
                Finding(code, Severity.ERROR, name, "amount is not a parseable decimal")
            )

    taxable = _amount(extraction.taxable_value)
    total = _amount(extraction.total_value)
    cgst = _zero_if_absent(extraction.cgst_amount)
    sgst = _zero_if_absent(extraction.sgst_amount)
    igst = _zero_if_absent(extraction.igst_amount)
    cess = _zero_if_absent(extraction.cess_amount)

    if taxable is not None and taxable < 0:
        findings.append(
            Finding(
                "NEGATIVE_TAXABLE_VALUE",
                Severity.ERROR,
                "taxable_value",
                "taxable value is negative; a credit note is a separate document type",
            )
        )

    if None in (taxable, total, cgst, sgst, igst, cess):
        return findings

    assert taxable is not None and total is not None
    assert cgst is not None and sgst is not None and igst is not None and cess is not None

    expected_total = taxable + cgst + sgst + igst + cess
    if abs(expected_total - total) > TOLERANCE:
        findings.append(
            Finding(
                "TOTAL_MISMATCH",
                Severity.ERROR,
                "total_value",
                f"taxable plus taxes is {expected_total}, invoice total is {total}",
            )
        )

    if cgst > 0 and sgst > 0 and abs(cgst - sgst) > TOLERANCE:
        findings.append(
            Finding(
                "CGST_SGST_ASYMMETRY",
                Severity.ERROR,
                "cgst_amount",
                f"central and state tax must be equal halves; got {cgst} and {sgst}",
            )
        )

    if taxable > 0:
        rate = ((cgst + sgst + igst) / taxable * 100).quantize(Decimal("0.01"))
        if rate.normalize() not in {r.normalize() for r in PLAUSIBLE_TAX_RATES}:
            findings.append(
                Finding(
                    "IMPLAUSIBLE_TAX_RATE",
                    Severity.WARNING,
                    "taxable_value",
                    f"implied combined tax rate of {rate}% is not a GST slab",
                )
            )

    return findings


def _check_tax_heads(extraction: InvoiceExtraction) -> list[Finding]:
    """Place of supply decides the tax head. Rule 46(l), (m) and (n).

    Supplier state equal to place of supply means an intra-state supply, taxed
    as CGST plus SGST. Otherwise it is inter-state and taxed as IGST.
    """
    supplier = extraction.supplier_gstin
    pos = extraction.place_of_supply_state_code
    if not supplier.is_present or not is_valid_gstin(str(supplier.value)):
        return []
    if not pos.is_present:
        return []

    cgst = _zero_if_absent(extraction.cgst_amount) or Decimal("0")
    sgst = _zero_if_absent(extraction.sgst_amount) or Decimal("0")
    igst = _zero_if_absent(extraction.igst_amount) or Decimal("0")

    findings = []
    if igst > 0 and (cgst > 0 or sgst > 0):
        findings.append(
            Finding(
                "MIXED_TAX_HEADS",
                Severity.ERROR,
                "igst_amount",
                "a supply is either intra-state or inter-state, never both",
            )
        )

    is_intra_state = str(supplier.value).strip()[:2] == str(pos.value).strip().zfill(2)
    if is_intra_state and igst > 0:
        findings.append(
            Finding(
                "WRONG_TAX_HEAD",
                Severity.ERROR,
                "igst_amount",
                "intra-state supply charged IGST",
            )
        )
    if not is_intra_state and (cgst > 0 or sgst > 0):
        findings.append(
            Finding(
                "WRONG_TAX_HEAD",
                Severity.ERROR,
                "cgst_amount",
                "inter-state supply charged CGST or SGST",
            )
        )
    return findings
